File: src/main/java/com/sisqueslabs/nexora/api/contexts/chat/transport/rest/dto/ChatCompletionRequestDto.java
package com.sisqueslabs.nexora.api.contexts.chat.transport.rest.dto;

import java.util.List;

import com.sisqueslabs.nexora.api.contexts.chat.application.command.createchatcompletion.CreateChatCompletionCommand;

/**
 * Mirrors the shape of the OpenAI API's POST /v1/chat/completions (only
 * the fields Nexora supports in v0: no streaming, no temperature/top_p/
 * etc yet).
 *
 * Purely a wire shape — no validation here. {@link #toCommandInput()} is
 * a mechanical, primitive-to-primitive mapping; the actual validation
 * happens once, in {@code CreateChatCompletionCommand}'s constructor.
 */
public record ChatCompletionRequestDto(String model, List<ChatCompletionMessageDto> messages) {

    public CreateChatCompletionCommand.Input toCommandInput() {
        List<CreateChatCompletionCommand.Input.MessageInput> messageInputs = messages == null
                ? List.of()
                : messages.stream()
                        .map(message -> new CreateChatCompletionCommand.Input.MessageInput(message.role(), message.content()))
                        .toList();

        return new CreateChatCompletionCommand.Input(model, messageInputs);
    }
}

File: src/contexts/chat/transport/rest/dtos/chat-completion-request.dto.ts
import { CreateChatCompletionCommandInput } from '@contexts/chat/application/commands/create-chat-completion/create-chat-completion.command';
import { ApiProperty } from '@nestjs/swagger';

import { ChatCompletionMessageDto } from './chat-completion-message.dto';

/**
 * Mirrors the shape of the OpenAI API's POST /v1/chat/completions (only
 * the fields Nexora supports in v0: no streaming, no temperature/top_p/
 * etc yet).
 *
 * Purely a wire shape — no validation here. toCommandInput() is a
 * mechanical, primitive-to-primitive mapping; the actual validation
 * happens once, in CreateChatCompletionCommand's constructor.
 */
export class ChatCompletionRequestDto {
  @ApiProperty({ example: 'nexora-mock-llama-3.1-8b' })
  model!: string;

  @ApiProperty({ type: [ChatCompletionMessageDto] })
  messages!: ChatCompletionMessageDto[];

  toCommandInput(): CreateChatCompletionCommandInput {
    return {
      model: this.model,
      messages: (this.messages ?? []).map((message) => ({
        role: message.role,
        content: message.content,
      })),
    };
  }
}
